File: C++/0297-serialize-and-deserialize-binary-tree.cpp
// Tags: Tree Design Amazon
// Time: O(n)
// Space: O(n)
class Codec {
public:
    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        ostringstream out;
        serializeHelper(root, out);
        // convert stream to string
        return out.str();
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        istringstream in {data};
        return deserializeHelper(in);
    }
private:
    // preorder traversal (push entries into stack)
    void serializeHelper(TreeNode* root, ostringstream &out) {
        if (root) {
            out << root->val << ' '; // add space
            serializeHelper(root->left, out);
            serializeHelper(root->right, out);
        } else {
            out << "# "; // add space
            }
        }
    // preorder traversal (pop stack entries from bottom side, same order)
    TreeNode *deserializeHelper(istringstream &in){
        string val {};
        in >> val;

        if (val == "#")
            return nullptr;

        TreeNode* node = new TreeNode(stoi(val));
        node->left = deserializeHelper(in);
        node->right = deserializeHelper(in);
        return node;
    }
};
/*
Encode in preorder fashion using insertion operator.
Similary using extraction operation we can retrieve back
the values in the same order.
*/

File: Python/0297-serialize-and-deserialize-binary-tree.py
# Time O(n)
# Space O(n)
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.

        :type root: TreeNode
        :rtype: str
        """
        def preOrder(node):
            if node:
                vals.append(str(node.val))
                preOrder(node.left)
                preOrder(node.right)
            else:
                vals.append('#')

        vals = []
        preOrder(root)
        return ' '.join(vals)

    def deserialize(self, data):
        """Decodes your encoded data to tree.

        :type data: str
        :rtype: TreeNode
        """
        def preOrder():
            val = next(vals)
            if val == '#':
                return None
            node = TreeNode(int(val))
            node.left = preOrder()
            node.right = preOrder()
            return node

        vals = iter(data.split())
        return preOrder()

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))
